### PR TITLE
Started implementing DER reader

### PR DESCRIPTION
**Pull request by masihyeganeh**
I could finally find some time to play with your code.
Here is a simple, ugly and maybe not correct implementation of DER reader.
It is easy to write a writer too, but first I need to check with you if I'm in right direction.
Please take a look at this and refactor it to fit better in your project.
I will then complete it.

I have a test that passes with this one, but some values are not right. Maybe it is related to SETs or order of fields.
Unfortunately I can't commit the test, but It would be great if you can write some tests for it.

I needed the SET feature, so I patched them locally, so it may not work without them.

Please help me implement it right.

Thanks

### CODE REVIEW

**Comment by masihyeganeh** [comment]
Also I did something weird to handle OPTIONAL. please let me know if it is OK or not

**Comment by kellerkindt** [comment]
I am currently in  the process of getting SET and SET OF support going - which is now possible since the `TagResolver` in cfb21015ccb255e7032b2dc1eca5f845e88ffe1b has landed.
After #19 has been merged, I will look into this in more detail.
I`ll leave one or two comments on the PR, nonetheless.

File: src/io/der/octet_aligned.rs
Comment (kellerkindt) [inline src/io/der/octet_aligned.rs:113]: As far as I can tell, DER/BER/CER are designed to read whole octets/bytes and then extract the information from these octets. I imagine something like (totally not tested!):

```rust
let octet = self.read_octet()?;
let bit8 = octet & 0x80 != 0; // required?
let bit7 = octet & 0b40 != 0; // required?
let class = Class:from((octet & 0xC0) >> 2);
```

This would allow us to drop the requirement on `OctetBuffer` for bit-oriented functions and only require simple byte-oriented functions.

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:160]: I don't understand why we are reading here in a `write_enumerated` function.
Since cfb21015ccb255e7032b2dc1eca5f845e88ffe1b `C::Tag` provides the `Tag` of the ENUMERATED instance.

**Review by kellerkindt — COMMENTED** [review]
Sorry, but just a hastily review.
Please also add unit tests for encoding/decoding messages.
You may add these as additional encoding tests to already existing uper test structs.

https://asn1.io/asn1playground/ Is a good place to verify the binary output.

File: src/syn/io/der.rs
Comment (masihyeganeh) [reply src/syn/io/der.rs:160]: It's not actually in `write_enumerated`, Github collapsed some lines here

File: src/io/der/octet_aligned.rs
Comment (masihyeganeh) [reply src/io/der/octet_aligned.rs:113]: You are right. I just tried to mimic your code for UPER. Bitwise operation is much better. I'll fix it

File: src/io/der/octet_aligned.rs
Comment (masihyeganeh) [reply src/io/der/octet_aligned.rs:88]: My bad, sorry

File: src/io/der/octet_aligned.rs
Comment (masihyeganeh) [reply src/io/der/octet_aligned.rs:6]: This was another mistake by me because I was trying to mimic your code.
I needed a non-perfect implementation to start correction with you.
I'll change it now

**Comment by masihyeganeh** [comment]
Ok, It's now a little better than before.
I just added two tests based on default example of [asn1 playground of ans1.io](https://asn1.io/asn1playground/).
It turn out that there is a lot of problems with tags and big numbers.

I also need to have access to details of field I am reading to be able to validate it. I think we need more data in `C`s.
I hope you can finalize SETs, Tags or anything that you are working on, then we implement this better.

for now, the `simple_der.rs` test passes, but `automatic_tags_der.rs` not.

File: src/io/der/octet_aligned.rs
Comment (masihyeganeh) [inline src/io/der/octet_aligned.rs:6]: Just saw there is an implementation for TAG, Type and Size in `asn1rs-model/src/model/mod.rs`.
I think it's better that I use them, then you provide them in `C` for each parser function, so I can validate them

Comment (masihyeganeh) [reply src/io/der/octet_aligned.rs:6]: My bad. You just implemented it in [cfb2101](https://github.com/kellerkindt/asn1rs/commit/cfb21015ccb255e7032b2dc1eca5f845e88ffe1b).
The code is a little fragmented for now, I think I need to wait for you to merge all of them, then we continue on top of that

File: src/syn/io/der.rs
Comment (kellerkindt) [reply src/syn/io/der.rs:159]: Sorry, my bad. You are right

File: src/io/der/octet_aligned.rs
Comment (kellerkindt) [reply src/io/der/octet_aligned.rs:6]: I just rebased the `experimental-set-setof` branch to include the changes of cfb2101.
I`ll probably merge it into master tomorrow (~20-ish hours from now).

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:201]: With `asn1rs::model::Type` this can be collapsed to
```rust
if !matches!(tag, Tag::Universal(n) if n != 16) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_SEQUENCE != tag {
    return Err(Error::InvalidType);
}
```

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:253]: (With `asn1rs::model::Type`) this can be collapsed to
```rust
if !matches!(tag, Tag::ContextSpecific(_)) {
    return Err(Error::InvalidType);
}
```

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:290]: With `asn1rs::model::Type` this can be collapsed to
```rust
if !matches!(tag, Tag::Universal(2)) || PC::Primitive != pc {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_INTEGER != tag || PC::Primitive != pc {
    return Err(Error::InvalidType);
}
```

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:313]: With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 12) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_UTF8_STRING != tag {
    return Err(Error::InvalidType);
}
```

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:335]: With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 22) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_IA5_STRING != tag {
    return Err(Error::InvalidType);
}
```

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:357]: With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 4) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_OCTET_STRING != tag {
    return Err(Error::InvalidType);
}
```

File: src/syn/io/der.rs
Comment (kellerkindt) [inline src/syn/io/der.rs:394]: With `asn1rs::model::Type` this can be collapsed to
```rust
if matches!(tag, Tag::Universal(n) if n != 1) || !matches!(len, LengthDefinitive(1)) {
    return Err(Error::InvalidType);
}
```
or more preferably
```rust
if Tag::DEFAULT_BOOLEAN != tag || !matches!(length, Length::Definitive(1)) {
    return Err(Error::InvalidType);
}
```

File: tests/automatic_tags_der.rs
Comment (kellerkindt) [inline tests/automatic_tags_der.rs:18]: As much as I like this example, for legal reasons I cannot accept their example in this repo

File: tests/simple_der.rs
Comment (kellerkindt) [inline tests/simple_der.rs:18]: As much as I like this example, for legal reasons I cannot accept their example in this repo

**Review by kellerkindt — CHANGES_REQUESTED** [review]
Added a few more notes. Do you mind adding a few more tests? For example, if you look in [` tests/basic_integer.rs: 49` at `fn test_uper_small()`](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_integer.rs#L49), it would be really nice to have equivalents for DER (like `fn test_der_small()`). The file [tests/basic_ia5string.rs](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs) is also a good place for inspiration.

Github Actions should also check this PR now automatically.

**Comment by masihyeganeh** [comment]
I think you need to rebase this branch on master, then we can do your changes

**Comment by kellerkindt** [comment]
The branch has already been merged on master :)

**Comment by masihyeganeh** [comment]
I mean the `der-support` branch. it needs to be rebased on master. Am I wrong?

**Comment by kellerkindt** [comment]
Oh, yes, you are right! Somehow I thought you this PR was for the `master` branch.

**Comment by masihyeganeh** [comment]
> Added a few more notes. Generally, I think you are on the right track 👍
> 
> Do you mind adding a few more tests? For example, if you look in [` tests/basic_integer.rs: 49` at `fn test_uper_small()`](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_integer.rs#L49), it would be really nice to have equivalents for DER (like `fn test_der_small()`). The file [tests/basic_ia5string.rs](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs) is also a good place for inspiration.
> 
> Github Actions should also check this PR now automatically.
> 
> PS: Please run rustfmt or cargo fmt, the CI won't accept it otherwise.

Ok, I rebased my code, changed things you said and made sure that tests passes.
It is far from complete, so please let the tests be these for now. I need to implement `write` to be able to test better.

For the next step, I want to do this:
Reader should should compare current tag with expected one and return `InvalidTag` error if they are not identical.
It may not be needed but I used this trick to implement optional.
For example if schema requires next tag to be an optional utf8string but instead it reads a number, that should return `InvalidTag` error, and the optional parser should receive that, reset read position and return None.
Please tell me if it is OK by you and if there is a better way

File: src/io/per/err.rs
Comment (kellerkindt) [inline src/io/per/err.rs:16]: I'like to see more debug info here. A `InvalidType(String)` in which the `String` contains something like `"Got unexpected tag {:?}"` or so.

File: tests/simple_der.rs
Comment (kellerkindt) [inline tests/simple_der.rs:24]: Thanks for changing the example. In the long term, I prefer this to have the same notation as in the [remaining crate](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs#L36)

File: tests/automatic_tags_der.rs
Comment (kellerkindt) [inline tests/automatic_tags_der.rs:24]: Thanks for changing the example. In the long term, I prefer this to have the same notation as in the [remaining crate](https://github.com/kellerkindt/asn1rs/blob/master/tests/basic_ia5string.rs#L36)

**Review by kellerkindt — COMMENTED** [review]
Nice, Thanks! So what's up next? I guess merging this into the `der-support` branch so you can focus on the writer impl which will be a follow-up PR?

Just as a side question: Is it valid DER to change the order of the elements while serializing? If so, the reader would need to be able to handle reordering the out-of-order elements in the long term? (If so, this could be a separate issue or a follow-up PR)

File: src/io/der/octet_aligned.rs
Comment (masihyeganeh) [reply src/io/der/octet_aligned.rs:72]: You are right. I'm trying hard to make it the same as your codes and it won't let me think about these simple things 😆 
I'll change it

File: src/io/per/err.rs
Comment (masihyeganeh) [reply src/io/per/err.rs:16]: Yup. Now that we have the `C::TAG`, it is better to change it to that

File: tests/automatic_tags_der.rs
Comment (masihyeganeh) [reply tests/automatic_tags_der.rs:24]: Sure. After implementing `Write`, I'll write separate tests for each data type like existing tests

File: tests/automatic_tags_der.rs
Comment (masihyeganeh) [inline tests/automatic_tags_der.rs:4]: Do you have a plan to implement this `AUTOMATIC TAGS`?

**Comment by masihyeganeh** [comment]
> Nice, Thanks! So what's up next? I guess merging this into the `der-support` branch so you can focus on the writer impl which will be a follow-up PR?
> 
> Just as a side question: Is it valid DER to change the order of the elements while serializing? If so, the reader would need to be able to handle reordering the out-of-order elements in the long term? (If so, this could be a separate issue or a follow-up PR)

Yes, I think it is OK for this PR to be merged (after implementing changes you wrote above) and continue on `Writer` in another PR.

I don't think it is valid in any of BER/CER/DER/XER/UPER/... to change order of elements (I may be wrong). I think it's only valid in SETs (I may be wrong about this too)

**Comment by masihyeganeh** [comment]
Ok, I think you can merge it now if you are OK with it.
I'll continue in another PR.
I think there is a problem with TAG implementation.
I'll work on that.

File: tests/automatic_tags_der.rs
Comment (kellerkindt) [reply tests/automatic_tags_der.rs:4]: To be honest, I didn't look into the Tag stuff too much, since it was never relevant for uPER. I am open to it being implemented.

**Review by kellerkindt — APPROVED** [review]
LFTM

**Pull request merged**
